### PR TITLE
Return HTTP 400 or 411 if Content-Length is invalid or empty

### PR DESCRIPTION
**Pull request by avtobiff**
Content-Length is defined to be an integer 0 or greater.

The current behaviour when Content-Length is not an integer is that Yaws immediately cuts the connection.

It is not totally clearly defined what to do when Content-Length is invalid, but responding with HTTP 400 Bad Request at least notifies the user of their error.

If POST is used and Content-Length is empty, respond with HTTP 411 Length Required.

### CODE REVIEW

File: src/yaws.erl [hunk 1]
```diff
@@ -2763,9 +2763,25 @@ http_collect_headers(CliSock, Req, H, SSL, Count) when Count < 1000 ->
...
             case H#headers.content_length of
                 undefined ->
-                    http_collect_headers(CliSock, Req,
-                                         H#headers{content_length = X},SSL,
-                                         Count+1);
+                    try
+                        %% assert content length is either empty or 0 or greater
+                        case X of
+                            "" -> ok;
+                            X -> true = 0 =< erlang:list_to_integer(X)
```
Comment (etnt) [inline src/yaws.erl:2770]: Parenthesis around expression?

File: src/yaws.erl [hunk 2]
```diff
@@ -2763,9 +2763,25 @@ http_collect_headers(CliSock, Req, H, SSL, Count) when Count < 1000 ->
         {ok, {http_header, _Num, 'Content-Length', _, X}} ->
             case H#headers.content_length of
                 undefined ->
-                    http_collect_headers(CliSock, Req,
-                                         H#headers{content_length = X},SSL,
-                                         Count+1);
+                    try
```
Comment (etnt) [inline src/yaws.erl:2766]: Perhaps avoid a 'try' here?

File: src/yaws.erl (hunk 1 above)
Comment (avtobiff) [reply src/yaws.erl]: Done.

**Comment by vinoski** [comment]
Is there any problem with this with `POST` requests that aren't required to have `Content-Length` headers, such as when chunked transfer is used? Given all checks have passed, including chunked transfer tests, seems like it's not an issue, but then I think that means the last sentence of the commit message isn't quite accurate.

**Comment by avtobiff** [comment]
I'll check with chunked transfer and POST.

**Comment by avtobiff** [comment]
Finally got time to rework this. It needed some changes to take
transfer encoding into account.

It should be ready for merge now.

**Comment by vinoski** [comment]
I merged thisto `master` manually. Thanks for contributing it!

**Pull request closed without merging**